// @nx/jest/src/generators/init/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jestInitGeneratorInternal = exports.jestInitGenerator = void 0;
const devkit_1 = require("@nx/devkit");
const update_package_scripts_1 = require("@nx/devkit/src/utils/update-package-scripts");
const plugin_1 = require("../../plugins/plugin");
const versions_1 = require("../../utils/versions");
function addPlugin(tree) {
    const nxJson = (0, devkit_1.readNxJson)(tree);
    nxJson.plugins ??= [];
    if (!nxJson.plugins.some((p) => typeof p === 'string'
        ? p === '@nx/jest/plugin'
        : p.plugin === '@nx/jest/plugin')) {
        nxJson.plugins.push({
            plugin: '@nx/jest/plugin',
            options: {
                targetName: 'test',
            },
        });
    }
    (0, devkit_1.updateNxJson)(tree, nxJson);
}
function updateProductionFileSet(tree) {
    const nxJson = (0, devkit_1.readNxJson)(tree);
    const productionFileSet = nxJson.namedInputs?.production;
    if (productionFileSet) {
        // This is one of the patterns in the default jest patterns
        productionFileSet.push(
        // Remove spec, test, and snapshots from the production fileset
        '!{projectRoot}/**/?(*.)+(spec|test).[jt]s?(x)?(.snap)', 
        // Remove tsconfig.spec.json
        '!{projectRoot}/tsconfig.spec.json', 
        // Remove jest.config.js/ts
        '!{projectRoot}/jest.config.[jt]s', 
        // Remove test-setup.js/ts
        // TODO(meeroslav) this should be standardized
        '!{projectRoot}/src/test-setup.[jt]s', '!{projectRoot}/test-setup.[jt]s');
        // Dedupe and set
        nxJson.namedInputs.production = Array.from(new Set(productionFileSet));
    }
    (0, devkit_1.updateNxJson)(tree, nxJson);
}
function addJestTargetDefaults(tree) {
    const nxJson = (0, devkit_1.readNxJson)(tree);
    nxJson.targetDefaults ??= {};
    nxJson.targetDefaults['@nx/jest:jest'] ??= {};
    const productionFileSet = nxJson.namedInputs?.production;
    nxJson.targetDefaults['@nx/jest:jest'].cache ??= true;
    // Test targets depend on all their project's sources + production sources of dependencies
    nxJson.targetDefaults['@nx/jest:jest'].inputs ??= [
        'default',
        productionFileSet ? '^production' : '^default',
        '{workspaceRoot}/jest.preset.js',
    ];
    nxJson.targetDefaults['@nx/jest:jest'].options ??= {
        passWithNoTests: true,
    };
    nxJson.targetDefaults['@nx/jest:jest'].configurations ??= {
        ci: {
            ci: true,
            codeCoverage: true,
        },
    };
    (0, devkit_1.updateNxJson)(tree, nxJson);
}
function updateDependencies(tree, options) {
    return (0, devkit_1.addDependenciesToPackageJson)(tree, {}, {
        '@nx/jest': versions_1.nxVersion,
        jest: versions_1.jestVersion,
    }, undefined, options.keepExistingVersions);
}
function jestInitGenerator(tree, options) {
    return jestInitGeneratorInternal(tree, { addPlugin: false, ...options });
}
exports.jestInitGenerator = jestInitGenerator;
async function jestInitGeneratorInternal(tree, options) {
    options.addPlugin ??= process.env.NX_ADD_PLUGINS !== 'false';
    if (!tree.exists('jest.preset.js')) {
        updateProductionFileSet(tree);
        if (options.addPlugin) {
            addPlugin(tree);
        }
        else {
            addJestTargetDefaults(tree);
        }
    }
    const tasks = [];
    if (!options.skipPackageJson) {
        tasks.push((0, devkit_1.removeDependenciesFromPackageJson)(tree, ['@nx/jest'], []));
        tasks.push(updateDependencies(tree, options));
    }
    if (options.updatePackageScripts) {
        await (0, update_package_scripts_1.updatePackageScripts)(tree, plugin_1.createNodes);
    }
    if (!options.skipFormat) {
        await (0, devkit_1.formatFiles)(tree);
    }
    return (0, devkit_1.runTasksInSerial)(...tasks);
}
exports.jestInitGeneratorInternal = jestInitGeneratorInternal;
exports.default = jestInitGenerator;
